File: LC#271-Encode_and_Decode_Strings/encode_and_decode_strings.py
# 271. Encode and Decode Strings
# Difficulty: Medium
# Topics: Array, String, Design
# https://leetcode.com/problems/encode-and-decode-strings/

from typing import List

class Solution:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string.
        """
        output = ""
        for word in strs:
            output += str(len(word)) + "#" + word
        
        return output
    
    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings.
        """
        result = []
        i = 0
        while i < len(s):
            j = i
            
            while s[j] != "#":
                j += 1

            # now we hit the delimeter
            word_len = int(s[i:j])
            start_word_indx = j + 1
            end_word_indx = start_word_indx + word_len
            word = s[start_word_indx:end_word_indx]
            result.append(word)

            i = end_word_indx
        return result
            
solution_instance = Solution()

# Test cases
case_1 = ["Hello","World"]
encoded_1 = solution_instance.encode(case_1)
decoded_1 = solution_instance.decode(encoded_1)

case_2 = [""]
encoded_2 = solution_instance.encode(case_2)
decoded_2 = solution_instance.decode(encoded_2)


assert case_1 == decoded_1
assert case_2 == decoded_2

print("All tests passed successfully!")

File: LC#271-Encode_and_Decode_Strings/encode_and_decode_strings.rs
/// 271. Encode and Decode Strings
/// Difficulty: Medium
/// Topics: Array, String, Design
/// https://leetcode.com/problems/encode-and-decode-strings/

fn main() {

    /// Encodes a list of strings to a single string.
    fn encode(strs: Vec<String>) -> String {
        let mut output = String::new();

        for word in strs {
            output.push_str(&word.len().to_string());
            output.push('#');
            output.push_str(&word);
        }

        output
    }
	
    /// Decodes a single string to a list of strings.
    fn decode(s: String) -> Vec<String> {
        let mut result = Vec::new();
        let mut i = 0;

        while i < s.len() {
            let mut j = i;

            while s.as_bytes()[j] != b'#' {
                j += 1;
            }

            // now we hit the delimiter
            let word_len: usize = s[i..j].parse().unwrap();

            let start_word_idx = j + 1;
            let end_word_idx = start_word_idx + word_len;

            let word = s[start_word_idx..end_word_idx].to_string();
            result.push(word);

            i = end_word_idx;
        }

        result
    }


    // Test cases
    let case_1 = vec!["Hello".to_string(), "World".to_string()];
    let case_1_encoded = encode(case_1.clone());
    let case_1_decoded = decode(case_1_encoded);

    let case_2 = vec!["".to_string()];
    let case_2_encoded = encode(case_2.clone());
    let case_2_decoded = decode(case_2_encoded);
    
    assert_eq!(case_1, case_1_decoded);
    assert_eq!(case_2, case_2_decoded);
    
    println!("All tests passed successfully!");
}